crates/events/src/lib.rs

```diff
@@ -214,14 +215,17 @@ thread_local! {
 ///             listener,
 ///         }
 ///     }
+///
+///     fn project_receiver(self: Pin<&mut Self>) -> Pin<&mut mpsc::UnboundedReceiver<()>> {
+///         unsafe { self.map_unchecked_mut(|s| &mut s.receiver) } 
+///     }
 /// }
 ///
 /// impl Stream for OnClick {
 ///     type Item = ();
-///     type Error = ();
 ///
-///     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-///         self.receiver.poll().map_err(|_| unreachable!())
+///     fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+///         self.project_receiver().poll_next(cx)
```

Comment:
This doesn't need `project_receiver` or unsafe code, it can just do this:

```rust
Pin::new(&mut self.receiver).poll_next(cx)
```

Reply:
I feel like I need to get more of an intuition for `Pin`s. Does this work because `UnboundedReceiver<()>` is `Unpin`?

Reply:
Yes, `Unpin` is similar to other auto-traits like `Sync` or `Send`: if all of the fields of your struct are `Unpin`, then the struct as a whole is `Unpin`. Almost all types in Rust are `Unpin`.

And if a struct is `Unpin` then you can safely access its fields (because of [this `DerefMut` impl](https://doc.rust-lang.org/nightly/std/pin/struct.Pin.html#impl-DerefMut)), and you can also safely use `Pin::new(&mut foo)` (because [`new` has an `Unpin` bound](https://doc.rust-lang.org/nightly/std/pin/struct.Pin.html#method.new)).

So the only time you need to use unsafe code is when one (or more) of your fields is *not* `Unpin`. This usually only happens in Future/Stream *combinators*, but not with leaf types like `OnClick`.

Or to put it another way: the burden of unsafe code is only for the creators of Future/Stream combinators, and not anybody else.